internal/gps/source.go

```diff
@@ -552,51 +524,122 @@ func (sg *sourceGateway) createSingleSourceCache() singleSourceCache {
 	return newMemoryCache()
 }
 
-func (sg *sourceGateway) require(ctx context.Context, wanted sourceState) (errState sourceState, err error) {
+// setUpSource tries to set src from choices and returns the initial sourceState,
+// otherwise it returns a slice of all errors. If force is true, then each
+// maybeSource may make a second attempt after removing the cache directory. testFn
+// is an optional check which may access src and return additional sourceState.
+func (sg *sourceGateway) setUpSource(ctx context.Context, force bool, testFn func(context.Context) (sourceState, error)) (sourceState, errorSlice) {
```

Comment:
if we're going to re-expose these methods, we need to be _diligent_ about not calling them from the outside. there was a rather huge mess back before i refactored to use the FSM, and much of it arose from mixing locks with intra-source object calls.

in fact, i'd probably be more comfortable if we separated these out by hiding the ones we expect `SourceMgr` itself to call (which should be mutex-protected) directly into an interface, and leaving the remainder unexposed.